src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs

```diff
@@ -57,59 +58,73 @@ public override int Read(Span<byte> buffer)
             if (_mode != CompressionMode.Decompress)
                 throw new InvalidOperationException(SR.BrotliStream_Compress_UnsupportedOperation);
             EnsureNotDisposed();
-            int totalWritten = 0;
 
-            OperationStatus lastResult = OperationStatus.DestinationTooSmall;
-            // We want to continue calling Decompress until we're either out of space for output or until Decompress indicates it is finished.
-            while (buffer.Length > 0 && lastResult != OperationStatus.Done)
+            int bytesWritten;
+            while (true)
             {
-                if (lastResult == OperationStatus.NeedMoreData)
+                // Decompress any data we may have in our buffer.
+                OperationStatus lastResult = _decoder.Decompress(new ReadOnlySpan<byte>(_buffer, _bufferOffset, _bufferCount), buffer, out int bytesConsumed, out bytesWritten);
```

Comment:
The next 40 lines or so are (I think) the same for the sync case and the async case. Should this be factored out into a separate method?

Reply:
Yup.  It's slightly cumbersome with the breaks, but not too bad.  Good suggestion.